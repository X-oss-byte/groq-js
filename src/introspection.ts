// Utilities for enabling introspection on query results

import type {GroqType} from './values/types'

export const introspectGroqType = Symbol('introspectGroqType')

export interface IntrospectableGroqType {
  [introspectGroqType]: GroqType
}
