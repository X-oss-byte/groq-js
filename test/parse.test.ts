import {parse} from '../src'

describe('Basic parsing', () => {
  test('Example query', () => {
    let query = `*[_type == "product"]{name}`
    let tree = parse(query)
    expect(tree).toMatchSnapshot()
  })
})

describe('Error reporting', () => {
  test('Query with syntax error', () => {
    let query = `*[_type == "]`
    try {
      parse(query)
    } catch (error) {
      expect(error.name).toBe('GroqSyntaxError')
      expect(error.position).toBe(13)
      expect(error.message).toBe('Syntax error in GROQ query at position 13')
    }
  })
})

describe('Delta-GROQ', () => {
  test('Supports before() and after()', () => {
    let query = `before().title == after().title`
    expect(() => parse(query)).toThrow()

    let tree = parse(query, {mode: 'delta'})
    expect(tree).toMatchSnapshot()
  })
})
