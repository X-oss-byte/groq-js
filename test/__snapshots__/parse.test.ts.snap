// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic parsing Example query 1`] = `
Object {
  "base": Object {
    "base": Object {
      "type": "Everything",
    },
    "expr": Object {
      "left": Object {
        "name": "_type",
        "type": "AccessAttribute",
      },
      "op": "==",
      "right": Object {
        "type": "Value",
        "value": "product",
      },
      "type": "OpCall",
    },
    "type": "Filter",
  },
  "expr": Object {
    "base": Object {
      "type": "This",
    },
    "expr": Object {
      "attributes": Array [
        Object {
          "name": "name",
          "type": "ObjectAttributeValue",
          "value": Object {
            "name": "name",
            "type": "AccessAttribute",
          },
        },
      ],
      "type": "Object",
    },
    "type": "Projection",
  },
  "type": "Map",
}
`;

exports[`Delta-GROQ Supports before() and after() 1`] = `
Object {
  "left": Object {
    "base": Object {
      "key": "before",
      "type": "Context",
    },
    "name": "title",
    "type": "AccessAttribute",
  },
  "op": "==",
  "right": Object {
    "base": Object {
      "key": "after",
      "type": "Context",
    },
    "name": "title",
    "type": "AccessAttribute",
  },
  "type": "OpCall",
}
`;
